File: src/test/java/integration/cardinstance/ability/CardInstanceAbilityFactoryTest.java
package integration.cardinstance.ability;

import static com.matag.cards.ability.type.AbilityType.HASTE;
import static com.matag.cards.ability.type.AbilityType.TRAMPLE;
import static java.util.Arrays.asList;
import static org.assertj.core.api.Assertions.assertThat;

import java.util.Optional;

import org.junit.Test;

import com.matag.game.cardinstance.ability.CardInstanceAbility;
import com.matag.game.cardinstance.ability.CardInstanceAbilityFactory;

public class CardInstanceAbilityFactoryTest {
  private CardInstanceAbilityFactory cardInstanceAbilityFactory = new CardInstanceAbilityFactory();

  @Test
  public void testAbilitiesFromParameters() {
    // Given
    var parameters = asList("+2/+2", "TRAMPLE", "DAMAGE:2", "HASTE");

    // When
    var abilities = cardInstanceAbilityFactory.abilitiesFromParameters(parameters);

    // Then
    assertThat(abilities).isEqualTo(asList(new CardInstanceAbility(TRAMPLE), new CardInstanceAbility(HASTE)));
  }

  @Test
  public void testNoAbilityFromParameter() {
    // Given
    var parameter = "+2/+2";

    // When
    var ability = cardInstanceAbilityFactory.abilityFromParameter(parameter);

    // Then
    assertThat(ability).isEmpty();
  }

  @Test
  public void testTrampleAbilityFromParameter() {
    // Given
    var parameter = "TRAMPLE";

    // When
    var ability = cardInstanceAbilityFactory.abilityFromParameter(parameter);

    // Then
    assertThat(ability).isEqualTo(Optional.of(new CardInstanceAbility(TRAMPLE)));
  }
}

File: src/test/java/integration/cardinstance/ability/CardInstanceAbilityFactoryTest.kt
package integration.cardinstance.ability

import com.matag.cards.ability.Ability
import com.matag.cards.ability.CardInstanceAbilityFactory
import com.matag.cards.ability.type.AbilityType
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import java.util.*

class CardInstanceAbilityFactoryTest {
    private val cardInstanceAbilityFactory = CardInstanceAbilityFactory()

    @Test
    fun testAbilitiesFromParameters() {
        // Given
        val parameters = mutableListOf<String?>("+2/+2", "TRAMPLE", "DAMAGE:2", "HASTE")

        // When
        val abilities = cardInstanceAbilityFactory.abilitiesFromParameters(parameters)

        // Then
        assertThat(abilities).isEqualTo(
            listOf(
                Ability(AbilityType.TRAMPLE),
                Ability(AbilityType.HASTE)
            )
        )
    }

    @Test
    fun testNoAbilityFromParameter() {
        // Given
        val parameter = "+2/+2"

        // When
        val ability = cardInstanceAbilityFactory.abilityFromParameter(parameter)

        // Then
        assertThat(ability).isEmpty()
    }

    @Test
    fun testTrampleAbilityFromParameter() {
        // Given
        val parameter = "TRAMPLE"

        // When
        val ability = cardInstanceAbilityFactory.abilityFromParameter(parameter)

        // Then
        assertThat(ability).isEqualTo(Optional.of(Ability(AbilityType.TRAMPLE)))
    }
}